site/book/07-effective-customizations/01-single-value-replacement.md

```diff
@@ -0,0 +1,137 @@
+## Scenario
+
+I have a single value replacement in my package. I don’t want package consumers 
+to look through all the yaml files to find the value I want them to set. It 
+seems easier to just create a parameter for this value and have the user look 
+at Kptfile for inputs.
+
+Example storage bucket:
+
+```yaml
+apiVersion: storage.cnrm.cloud.google.com/v1beta1
+kind: StorageBucket
+metadata:
+  name: my-bucket # kpt-set: ${project-id}-${name}
+  namespace: ns-test # kpt-set: ${namespace}
+  annotations:
+    cnrm.cloud.google.com/force-destroy: "false"
+    cnrm.cloud.google.com/project-id: my-project # kpt-set: ${project-id}
+spec:
+  storageClass: standard # kpt-set: ${storage-class}
+  uniformBucketLevelAccess: true
+  versioning:
+    enabled: false
+```
+
+The corresponding Kptfile:
+
+```yaml
+apiVersion: kpt.dev/v1
+kind: Kptfile
+metadata:
+  name: bucket
+info:
+  description: A Google Cloud Storage bucket
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.2
+      configMap:
+        name: todo-bucket-name
+        namespace: todo-namespace
+        project-id: todo-project-id
+        storage-class: standard
+```
+
+
+## Problems
+
+1. With package popularity the single values inevitably expand to provide a 
+facade to a large portion of the data.  That defeats the purpose of minimizing 
+the cognitive load.  With this small example almost half of the StorageBucket configuration is now covered with parameters.
+1. Some values like resource names are used as references so setting them in 
+one place needs to trigger updates in all the places where they are referenced.
+1. If additional resources that have similar values are added to the package 
+new string replacements need to be added.  In this case everything will need
+to also be marked up with project ID and namespace.
+1. If a package is used as a sub-package the string replacement parameters need 
+to be surfaced to the parent package and if the parent package already expects 
+some values to be set and the parameters do not exist, the sub-package needs to 
+be updated.
+
+## Solutions:
+
+1. kpt allows the user to edit a particular value directly in the configuration 
+data and will handle upstream merge.  When [editing the yaml] directly the 
+consumers are not confined to the parameters that the package author has 
+provided.  kpt has made an investment into upstream merge that allows 
+[updating a package] that has been changed with a text editor.  In this case 
```

Comment:
nit: kpt has made an investment into upstream merge that allows.....
`investment` word feels a bit odd here. An alternative suggestion:
`kpt pkg update` [link to the pkg-update] merges the local edits made by consumer seamlessly with the changes in the upstream pkg made by publisher.

Reply:
fixed, thank you!